xdsl/transforms/lower_affine.py

```diff
@@ -0,0 +1,148 @@
+from xdsl.dialects import affine, arith, builtin, memref, scf
+from xdsl.ir import MLContext, SSAValue
+from xdsl.ir.affine import AffineConstantExpr
+from xdsl.ir.affine.affine_expr import (
+    AffineBinaryOpExpr,
+    AffineBinaryOpKind,
+    AffineDimExpr,
+    AffineSymExpr,
+)
+from xdsl.ir.core import Operation
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+
+def affine_expr_ops(
+    expr: affine.AffineExpr,
+    dims: list[SSAValue],
+    symbols: list[SSAValue],
+) -> tuple[list[Operation], SSAValue]:
+    """
+    Returns the operations that evaluate the affine expression when given input SSA
+    values, along with the SSAValue representing the result.
+    """
+    if isinstance(expr, AffineConstantExpr):
+        constant = arith.Constant(builtin.IntegerAttr.from_index_int_value(expr.value))
+        return [constant], constant.result
+
+    if isinstance(expr, AffineDimExpr):
+        return [], dims[expr.position]
+    if isinstance(expr, AffineSymExpr):
+        return [], symbols[expr.position]
+
+    if isinstance(expr, AffineBinaryOpExpr):
+        lhs_ops, lhs_val = affine_expr_ops(expr.lhs, dims, symbols)
+        rhs_ops, rhs_val = affine_expr_ops(expr.rhs, dims, symbols)
+
+        match expr.kind:
+            case AffineBinaryOpKind.Add:
+                op = arith.Addi(lhs_val, rhs_val)
+            case AffineBinaryOpKind.Mul:
+                op = arith.Muli(lhs_val, rhs_val)
+            case AffineBinaryOpKind.Mod:
+                op = arith.RemSI(lhs_val, rhs_val)
+            case AffineBinaryOpKind.FloorDiv:
+                op = arith.FloorDivSI(lhs_val, rhs_val)
+            case AffineBinaryOpKind.CeilDiv:
+                op = arith.CeilDivSI(lhs_val, rhs_val)
+
+        return [*lhs_ops, *rhs_ops, op], op.result
+
+    raise ValueError("Unreachable")
+
+
+def insert_affine_map_ops(
+    map: affine.AffineMapAttr | None,
+    dims: list[SSAValue],
+    symbols: list[SSAValue],
+) -> tuple[list[Operation], list[SSAValue]]:
+    """
+    Returns operations that evaluate the affine map when given input SSA values and the
+    resulting indices.
+    """
+    if map is None:
+        ops = []
+        indices = dims
+    else:
+        ops: list[Operation] = []
```

Comment:
Why define ops to times and not just do it like this:

```suggestion
     ops: list[Operation] = []
     if map is None:
        indices = dims
    else:
```

Reply:
In a language that isn't python those might have slightly different semantics but this is not the case here :)

Reply:
I just prefer reading the lower thing, but I also don't have preferences.

I guess I am just more used to block-scopes than function-scopes from reading c++...

Reply:
Done :)